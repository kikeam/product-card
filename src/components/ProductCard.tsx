import styles from '../styles/styles.module.css';

import { useProduct } from '../hooks/useProduct';
import React, { createContext } from 'react';
import {
  Product,
  onChangeArgs,
  initialValues,
  ProductCardHandlers,
} from '../interfaces/interfaces';

import { ProductContextProps } from '../interfaces/interfaces';

export interface Props {
  children: (args: ProductCardHandlers) => JSX.Element;
  product: Product;
  className?: string;
  style?: React.CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: initialValues;
}

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const { counter, increaseBy, maxCount, isMaxReached, reset } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });
  return (
    <Provider value={{ increaseBy, counter, product, maxCount }}>
      <div style={style} className={`${styles.productCard} ${className}`}>
        {children({
          count: counter,
          isMaxReached,
          maxQuantity: initialValues?.maxQuantity,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
