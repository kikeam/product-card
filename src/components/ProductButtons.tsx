import React, { useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface Props {
  className?: string;
}

export const ProductButtons = ({ className }: Props) => {
  const { counter, increaseBy, maxCount } = useContext(ProductContext);

  const isMaxReached = useCallback(() => !!maxCount && counter === maxCount, [
    counter,
    maxCount,
  ]);

  return (
    <div className={`${styles.buttonsContainer} ${className}`}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        className={`${styles.buttonAdd} `}
        onClick={() => increaseBy(1)}
        disabled={isMaxReached()}
      >
        +
      </button>
    </div>
  );
};
