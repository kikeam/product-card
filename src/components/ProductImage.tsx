import React, { useContext } from 'react';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface Props {
  img?: string;
  className?: string;
}

export const ProductImage = ({ img, className }: Props) => {
  const { product } = useContext(ProductContext);
  const image = img || product.img || noImage;
  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={image}
      alt="Cafe"
    />
  );
};
