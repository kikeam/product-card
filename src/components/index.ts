import { ProductCard as ProductCardComponent } from './ProductCard';
import { ProductTitle } from './ProductTitle';
import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductCardComponentProps } from '../interfaces/interfaces';
export { ProductButtons } from './ProductButtons';
export { ProductTitle } from './ProductTitle';
export { ProductImage } from './ProductImage';
//export { ProductCard } from './ProductCard';


export const ProductCard:ProductCardComponentProps =Object.assign(ProductCardComponent, {
    Title: ProductTitle,
    Buttons: ProductButtons,
    Image: ProductImage
});

export default ProductCard;