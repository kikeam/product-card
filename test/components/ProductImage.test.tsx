import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components/ProductCard';
import { ProductImage } from '../../src/components/ProductImage';
import { product2 } from '../data/products';

describe('Tests on ProductImage', () => {
  test('should display component correctly with personalized image', () => {
    const image = '../../src/assets/no-image.jpg';
    const wrapper = renderer.create(<ProductImage img={image} />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  test('should display component correctly with default image', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
