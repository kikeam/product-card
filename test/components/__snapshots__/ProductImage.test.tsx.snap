// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on ProductImage should display component correctly with default image 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Cafe"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`Tests on ProductImage should display component correctly with personalized image 1`] = `
<img
  alt="Cafe"
  className="productImg undefined"
  src="../../src/assets/no-image.jpg"
/>
`;
